### PR TITLE
Unable to stop CollectClosables thread. Resolve #758

### PR DESCRIPTION
**Pull request by DanielCharczynski**
module-name: util-core

Problem

No possibility to stop _CollectClosables_ thread.
Problem described here https://github.com/twitter/finagle/issues/758

Solution

Created public method `Closable.stopCollectClosablesThread` that stops _CollectClosables_ thread

Result

New API that makes possible to stop _CollectClosables_ thread.

### CODE REVIEW

**Comment by DanielCharczynski** [comment]
I have signed Contributor License Agreement five times but it looks like it is no visible here....

**Comment by mosesn** [comment]
This patch seems reasonable to me.  Can you follow the guidelines in CONTRIBUTING.md about your commit message / github PR description? LGTM otherwise.

**Comment by DanielCharczynski** [comment]
Done

**Comment by mosesn** [comment]
Great, thanks!  We'll make sure it passes our internal tests too and then merge it in.

File: util-core/src/main/scala/com/twitter/util/Closable.scala [hunk 1]
```diff
@@ -45,6 +45,14 @@ abstract class AbstractClosable extends Closable
  */
 object Closable {
   private[this] val logger = Logger.getLogger("")
+  private[this] var collectorThreadEnabled = true
```
Comment (ryanoneill) [inline util-core/src/main/scala/com/twitter/util/Closable.scala]: What do you think about making this volatile? I would expect the callers of `stopCollectClosablesThread` to not be the "CollectClosables" thread itself, and so there may be memory visibility issues here.

File: util-core/src/main/scala/com/twitter/util/Closable.scala [hunk 2]
```diff
@@ -45,6 +45,14 @@ abstract class AbstractClosable extends Closable
...
+  
+  
+  /** Stop CollectClosables thread. */
+  def stopCollectClosablesThread() {
+	collectorThreadEnabled = false
+	collectorThread.interrupt()
```
Comment (ryanoneill) [inline util-core/src/main/scala/com/twitter/util/Closable.scala]: Calling `interrupt` can throw a `SecurityException`. It _shouldn't_ in our case, but perhaps we should catch the exception here, log and just eat it.

File: util-core/src/main/scala/com/twitter/util/Closable.scala [hunk 3]
```diff
@@ -45,6 +45,14 @@ abstract class AbstractClosable extends Closable
  */
 object Closable {
   private[this] val logger = Logger.getLogger("")
+  private[this] var collectorThreadEnabled = true
+  
+  
+  /** Stop CollectClosables thread. */
+  def stopCollectClosablesThread() {
+	collectorThreadEnabled = false
```
Comment (ryanoneill) [inline util-core/src/main/scala/com/twitter/util/Closable.scala]: We should perhaps protect against calling this multiple times.

File: util-core/src/main/scala/com/twitter/util/Closable.scala (hunk 1 above)
Comment (ryanoneill) [reply util-core/src/main/scala/com/twitter/util/Closable.scala]: Could also make it an AtomicBoolean.

Comment (DanielCharczynski) [reply util-core/src/main/scala/com/twitter/util/Closable.scala]: AtomicBoolean already uses 'volatile' variable

File: util-core/src/main/scala/com/twitter/util/Closable.scala (hunk 3 above)
Comment (DanielCharczynski) [reply util-core/src/main/scala/com/twitter/util/Closable.scala]: OK

File: util-core/src/main/scala/com/twitter/util/Closable.scala (hunk 2 above)
Comment (DanielCharczynski) [reply util-core/src/main/scala/com/twitter/util/Closable.scala]: Done

**Comment by jyanJing** [comment]
Thank you so much for working on it @DanielCharczynski , the change is merged in [a826099](https://github.com/twitter/util/commit/a8260998518a9efc32ece0c5114694674c63072c)

**Pull request closed without merging**